src/high-level/vector.lisp

```diff
@@ -0,0 +1,155 @@
+       (defmethod copy-tensor ((v ,name) &rest args)
+         (declare (ignore args))
+         (let ((new-v (,copy-sym v)))
+           (setf (,storage-sym new-v)
+                 (make-array (vector-size v) :element-type (element-type v)))
+           new-v))
+
+       (defmethod deep-copy-tensor ((v ,name) &rest args)
+         (declare (ignore args))
+         (let ((new-v (,copy-sym v)))
+           (setf (,storage-sym new-v)
+                 (copy-seq (,storage-sym v)))
+           new-v))
+
+       (defmethod tref ((vector ,name) &rest pos)
+         (policy-cond:with-expectations (> speed safety)
+             ((assertion (valid-index-p pos (list (vector-size vector)))))
+           (aref (,storage-sym vector) (first pos))))
+
+       (defmethod (setf tref) (new-value (vector ,name) &rest pos)
+         (policy-cond:with-expectations (> speed safety)
+             ((assertion (valid-index-p pos (list (vector-size vector)))))
+           (setf (aref (,storage-sym vector) (first pos)) new-value))))))
+
+
+(defun pprint-vector (stream vector)
+  "Pretty-print a vector VECTOR to the stream STREAM."
+  (flet ((print-real (x)
+           (format stream "~6,3f" x))
+         (print-complex (z)
+           (format stream "~6,3f ~:[+~;-~]~6,3fj"
+                   (realpart z)
+                   (minusp (imagpart z))
+                   (abs (imagpart z))))
+         (print-int (x)
+           (format stream "~3d" x)))
+    (let* ((size (size vector))
+           (type (element-type vector))
+           (print-entry
+             (cond
+               ((subtypep type 'complex) #'print-complex)
+               ((subtypep type 'integer) #'print-int)
+               (t #'print-real))))
+      (pprint-logical-block (stream nil)
+        (print-unreadable-object (vector stream :type t)
+          (format stream "(~D):" size)
+          (dotimes (e size)
+            (pprint-newline :mandatory stream)
+            (funcall print-entry (tref vector e))))))))
+
+(set-pprint-dispatch 'vector 'pprint-vector)
+
+;;; Required abstract-tensor methods
+
+(defmethod order ((vector vector))
+  (declare (ignore vector))
+  1)
+
+(defmethod shape ((vector vector))
+  (list (vector-size vector)))
+
+(defmethod (setf shape) (new-value (vector vector))
+  (policy-cond:with-expectations (> speed safety)
+      ((type shape new-value)
+       (assertion (cl:= 1 (length new-value))))
+    (setf (vector-size vector) (first new-value))))
+
+(defgeneric dot (vector1 vector2)
+  (:documentation "Compute the dot product of two vectors")
+  (:method ((vector1 vector) (vector2 vector))
+    (policy-cond:with-expectations (> speed safety)
+        ((assertion (cl:= (size vector1) (size vector2))))
+      (loop :for i :below (size vector1)
+            :sum (* (tref vector1 i) (tref vector2 i))))))
+
+(defgeneric norm (vector &optional p)
+  (:documentation "Compute the norm of a vector")
+  (:method ((vector vector) &optional (p 2))
+    (expt (reduce #'+
```

Comment:
What about the case p=+∞ ?

Reply:
Added in 015e273. You can call `(norm vect :infinity)` or `(norm vect :neg-inf)`